src/libraries/System.Security.Claims/src/System/Security/Claims/ClaimsIdentity.cs

```diff
@@ -698,14 +699,39 @@ public virtual bool HasClaim(string type, string value)
 
             foreach (Claim claim in Claims)
             {
-                if (claim != null
-                        && string.Equals(claim.Type, type, StringComparison.OrdinalIgnoreCase)
-                        && string.Equals(claim.Value, value, StringComparison.Ordinal))
+                if (string.Equals(claim.Type, type, StringComparison.OrdinalIgnoreCase))
                 {
-                    return true;
+                    if (claim.Value != null)
+                    {
+                        if (string.Equals(claim.Value, value, StringComparison.Ordinal))
+                        {
+                            return true;
+                        }
+                        else if (claim.Value.Trim().StartsWith("[") && claim.Value.Trim().EndsWith("]"))
+                        {
+                            try
+                            {
+                                string[]? claimValues = JsonSerializer.Deserialize<string[]>(claim.Value.ToString());
+
+                                if (claimValues != null)
+                                {
+                                    foreach (string claimValue in claimValues)
+                                    {
+                                        if (string.Equals(claimValue, value, StringComparison.Ordinal))
+                                        {
+                                            return true;
+                                        }
+                                    }
+                                }
+                            }
+                            catch (JsonException)
+                            {
+                                return false;
```

Comment:
Shouldn't it compare the other claims, instead of returning false right away?

Reply:
A yes, improved also the test